relay-general/src/processor/attrs.rs

```diff
@@ -212,6 +213,61 @@ pub struct FieldAttrs {
     pub retain: bool,
 }
 
+/// A set of characters allowed or denied for a (string) field.
+///
+/// Note that this field is generated in the derive, it can't be constructed easily in tests.
+#[derive(Clone, Copy)]
+pub struct CharacterSet {
+    /// Generated in derive for performance. Can be left out when set is created manually.
+    pub is_match: Option<fn(char) -> bool>,
+    /// A set of ranges that are allowed/denied within the character set
+    pub ranges: &'static [RangeInclusive<char>],
+    /// Whether the character set is inverted
+    pub is_negative: bool,
+}
+
+impl CharacterSet {
+    pub fn allow(ranges: &'static [RangeInclusive<char>]) -> Self {
+        CharacterSet {
+            ranges,
+            is_match: None,
+            is_negative: false,
+        }
+    }
+
+    pub fn deny(ranges: &'static [RangeInclusive<char>]) -> Self {
+        CharacterSet {
+            ranges,
+            is_match: None,
+            is_negative: true,
+        }
+    }
+
+    #[inline]
+    pub fn is_match(&self, c: char) -> bool {
```

Comment:
This is a slightly confusing function to me at the moment. If I'm reading the code correctly, then `is_match` means `is_invalid`. Would it make more sense to just generate a function `is_valid`?

Reply:
`is_match` means `is_valid`. There is a double-negation with `is_negative`, which is false if the character set is inclusive. I can remove this function though if you're suggesting to not make this constructible outside of derive (which tbh I am not sure of, it does mean you can't impl traits manually